### PR TITLE
CPLAT-14830 CPLAT-15336 MUI migration codemod

### PR DESCRIPTION
**Pull request by greglittlefield-wf**
## Motivation
We need some codemods that migrate from one component library (our internal web_skin_dart library) to another (our internal react_material_ui library).

We also need utilities and patterns for migrating component usages.

## Changes
Some changes were added in separate PRs. TODO Greg add those

- Add base utilities component usage codemods
  - Add `ComponentUsageMigrator`
  - Update `FluentComponentUsage` with convenience utilities for inspecting cascades and children, as well as for dealing with static information
  - Add utilities for getting resolved codemod `FileContexts` with package import supports within tests
- Add suggestors for that migrate the WSD `Button` (and `FormSubmitInput`/`FormResetInput`), `ButtonGroup`, and `ButtonToolbar` components to MUI alternatives
- Add suggestors to add MUI imports and remove unused WSD imports
- Add mui_migration codemod executable
- Add tests

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Client Platform member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react_codemod/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react_codemod/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/util/wsd_util.dart
Comment (greglittlefield-wf) [reply lib/src/util/wsd_util.dart:181]: It might come in handy in future codemods, and it's already tested, so I'd vote to keep it in.

File: lib/src/util/ignore_info.dart
Comment (greglittlefield-wf) [reply lib/src/util/ignore_info.dart]: Good call-out. This was a magic string signifying that all codes would be ignored, which really wasn't a clean or robust solution.

I reworked this file to not make use of it anymore in 1b4e347ddc66b706f27fdb16ec5595a88e4dcefd

File: lib/src/element_type_helpers.dart
Comment (greglittlefield-wf) [reply lib/src/element_type_helpers.dart]: Good catch! Updated to reference that function

**Comment by greglittlefield-wf** [comment]
@sydneyjodon-wk All your feedback has been addressed!

**Review by aaronlademann-wf — APPROVED** [review]
+1

File: test/util/element_type_helpers_test.dart
Comment (sydneyjodon-wk) [inline test/util/element_type_helpers_test.dart:150]:
```suggestion
          expect(parameterType('element').isPropsClass, isFalse);
          expect(parameterType('map').isPropsClass, isFalse);
          expect(parameterType('customClass').isPropsClass, isFalse);
        });

        test('returns false for dynamic', () {
          expect(parameterType('dynamic_').isPropsClass, isFalse);
```

Comment (greglittlefield-wf) [reply test/util/element_type_helpers_test.dart:150]: Oof, good catch!

**Review by sydneyjodon-wk — APPROVED** [review]

**Review by joebingham-wk — COMMENTED** [review]
+1 to the docs!!

**Review by sydneyjodon-wk — DISMISSED** [review]

File: lib/src/util/component_usage.dart
Comment (kealjones-wk) [inline lib/src/util/component_usage.dart:244]: This is also kinda in relation to `ExpressionComponentChild` what happens in situations like
```
Dom.div()(
'foo',
['bar'],
)
```
what would bar get marked as in this situation? and does `isVariadic` apply?

File: lib/src/util/component_usage_migrator.dart
Comment (kealjones-wk) [inline lib/src/util/component_usage_migrator.dart:39]:
```suggestion
/// Throws if any file cannot be resolved or if any of the OverReact component
```

File: lib/src/util/component_usage_migrator.dart
Comment (kealjones-wk) [inline lib/src/util/component_usage_migrator.dart:529]: The example code block isn't terminated
```suggestion
  /// ```dart
  /// // Flagging "id" for this usage...
  /// (Dom.div()..id = '')();
  ///
  /// // ...results in this...
  /// (Dom.div()
  ///   // FIXME ...
  ///   ..id = ''
  /// )();
  ///
  /// // ...and not this...
  /// (Dom.div() // FIXME ...
  ///   ..id = ''
  /// )();
  /// ```
  void yieldBuilderMemberFixmePatch(
```

File: lib/src/util/component_usage_migrator.dart
Comment (kealjones-wk) [inline lib/src/util/component_usage_migrator.dart:567]: Example codeblock isn't terminated
```suggestion
  /// For example:
  /// ```dart
  /// // Flagging the single child for this usage...
  /// Dom.div()('child');
  ///
  /// // ...results in this...
  /// Dom.div()(
  ///   // FIXME
  ///   'child'
  /// );
  ///
  /// // ...and not this...
  /// Dom.div()( // FIXME ...
  ///   'child'
  /// );
  /// ```
  void yieldChildFixmePatch(ComponentChild child, String message) {
```

File: lib/src/util/component_usage_migrator.dart
Comment (kealjones-wk) [inline lib/src/util/component_usage_migrator.dart:758]: hehe, so meta

File: lib/src/util/package_util.dart
Comment (kealjones-wk) [inline lib/src/util/package_util.dart:64]: lol

File: lib/src/util/package_util.dart
Comment (kealjones-wk) [inline lib/src/util/package_util.dart:70]: ` if (exitCode == 69)`...`inheritStd` uh oh

**Review by kealjones-wk — CHANGES_REQUESTED** [review]
Honestly nothing in my review is vital, feel free to dismiss, I haven't reviewed the tests yet, but i need lunch.

**Review by aaronlademann-wf — DISMISSED** [review]
+10

Changes produced by this codemod for Button, ButtonGroup, ButtonToolbar et. al. can be previewed [here](https://sourcegraph.wk-dev.wdesk.org/organizations/Workiva/batch-changes/apply/QmF0Y2hTcGVjOiI1VWFrV0luS2ZuWiI=)

**Review by sydneyjodon-wk — APPROVED** [review]

File: lib/src/executables/mui_migration.dart
Comment (greglittlefield-wf) [reply lib/src/executables/mui_migration.dart:53]: It looks like `negatable` actually defaults to `true`

File: lib/src/util.dart
Comment (greglittlefield-wf) [reply lib/src/util.dart:432]: Yeah, there's nowhere that says that generator functions have to return non-empty iterables/streams.

I think the idea for other functions is that if you're not explicitly returning a value, then you probably missed a return statement by mistake. See https://github.com/dart-lang/sdk/issues/35710 for more info.

File: lib/src/util/component_usage.dart
Comment (greglittlefield-wf) [reply lib/src/util/component_usage.dart:244]: In that case,  you'd have two `ExpressionComponentChild`s, one for `'foo'` and one for `['bar']`

**Review by sydneyjodon-wk — APPROVED** [review]

**Review by aaronlademann-wf — APPROVED** [review]
+1 / +10

**Comment by aaronlademann-wf** [comment]
@Workiva/release-management-pp

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM

**Pull request merged**
